skypy/halo/mass.py

```diff
@@ -314,3 +320,161 @@ def _sigma_squared(M, k, Pk, growth_function, cosmology):
+    array([1.50205889e-07, 1.68169756e-08])
+
+    If we now choose the model given by equation 4 in [1]:
+
+    >>> params_vale = (0.5, 1.9, 1.0, 1.0e9)
+    >>> mass.subhalo_mass_function(m, 1.0e12, params_vale, model='Vale')
+    array([2.78569506e-08, 3.11885015e-09])
+
+    References
+    ----------
+    .. [1] Vale, A. and Ostriker, J.P. (2005), arXiv: astro-ph/0511816.
+    '''
+    if model == 'Vale':
+        subhalo_fraction0, alpha, beta, mcut = params
+        A = _subhalo_amplitude(halo_mass, params)
+    else:
+        A, alpha, beta = params
+
+    x = M / (beta * halo_mass)
+
+    return A * np.power(x, -alpha) * np.exp(-x) / (beta * halo_mass)
+
+
+def _subhalo_amplitude(M, params):
+    subhalo_fraction0, alpha, beta, mcut = params
+    x_cut = mcut / (beta * M)
+
+    return (subhalo_fraction0 / beta) / \
+        gamma(2.0 - alpha) * (1.0 - gammainc(2.0 - alpha, x_cut))
+
+
+def subhalo_mass_sampler(m_min, m_max, resolution,
+                         halo_mass, params, model=None):
+    r'''Subhalo mass sampler.
+    This function samples subhaloes from their mass function, given the mass
+    of the parent halo. Refer to equation 3 in [1]_.
+
+    Parameters
+    -----------
+    m_min, m_max : array_like
+        Lower and upper bounds for the random variable m.
+    resolution: int
+        Resolution of the inverse transform sampling spline.
+    M : (nm,) array_like
+        Array for the subhalo mass, in units of solar mass.
+    halo_mass : (nm, ) array_like
+        The mass of the halo parent, in units of solar mass.
+    params: tuple
+        List of parameters that determines the subhalo Schechter function.
+        If model is `None`, the parameters are `(A, alpha, beta)`.
+        If model is `'Vale'`, the parameters are
+        `(subhalo_fraction0, alpha, beta, mcut)`
+        and the amplitude is defined by equation 4 in [1].
+    model: str
+        Model for the amplitude of the subhalo mass function. Default is None.
+
+    Returns
+    --------
+    sample: List
+        Samples drawn from the mass function, in units of solar masses.
+
+    Examples
+    ---------
+    >>> import numpy as np
+    >>> from skypy.halo import mass
+
+    This example samples from the subhalo mass function given a parent halo
+    of mass :math:`10^12 M_{Sun}`:
+
+    >>> params_model = (0.3, 1.9, 1.0)
+    >>> sh = mass.subhalo_mass_sampler(1e9, 1e10, 100, 1.0e12, params_model)
+
+    If we now choose the model given by equation 4 in [1]:
+
+    >>> params_vale = (0.5, 1.9, 1.0, 1.0e9)
+    >>> sh = mass.subhalo_mass_sampler(1e9, 1e10, 100, 1.0e12, params_vale,
+    ...      model='Vale')
+
+
+    References
+    ----------
+    .. [1] Vale, A. and Ostriker, J.P. (2005), arXiv: astro-ph/0511816.
+    '''
+
+    if model == 'Vale':
+        subhalo_fraction0, alpha, beta, mcut = params
+        A = _subhalo_amplitude(halo_mass, params)
+    else:
+        A, alpha, beta = params
```

Comment:
For now lets just implement the V&O subhalo model so make `subhalo_fraction0, alpha, beta, mcut` arguments of the function instead of using `params`